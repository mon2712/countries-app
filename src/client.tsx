import { ApolloClient, InMemoryCache } from '@apollo/client';

const client = new ApolloClient({
    uri: 'https://countries-274616.ew.r.appspot.com/',
    cache: new InMemoryCache(),
});

export default client;
