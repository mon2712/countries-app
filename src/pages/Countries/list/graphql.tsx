import { gql } from '@apollo/client';

import { fragments } from './fragments';

export const COUNTRIES = gql`
    query GetCountriesList($first: Int, $offset: Int, $filter: _CountryFilter) {
        Country(first: $first, offset: $offset, filter: $filter) {
            ...CountryAttributes
        }
        Currency {
            _id
            name
            code
        }
        Language {
            _id
            name
        }
        Region {
            _id
            name
        }
    }
    ${fragments.country}
`;
