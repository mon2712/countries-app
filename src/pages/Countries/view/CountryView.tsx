import React from 'react';
import { useQuery } from '@apollo/client';
import { withRouter } from 'react-router-dom';

import { COUNTRY_DATA } from './graphql';
import { CountryCountainer, GoBackLink } from './CountryView.styles';
import { CountryData, CountryVars } from './index';
import Page from 'components/Page';
import Loader from 'components/Loader';
import { Typography } from '@material-ui/core';
import { Grid } from '@material-ui/core';
import { Alert, AlertTitle } from '@material-ui/lab';

const CountryView = (props: any) => {
    const { match } = props;
    const { loading, error, data } = useQuery<CountryData, CountryVars>(
        COUNTRY_DATA,
        {
            variables: {
                _id: match.params.Countries_id,
            },
        },
    );

    if (error) {
        <Alert severity="error">
            <AlertTitle>Error</AlertTitle>
            {error.message}
        </Alert>;
    }

    if (loading) {
        return (
            <Page title="Cargando">
                <Typography variant="h1">Loading...</Typography>
                <Loader />
            </Page>
        );
    }

    if (!data || !data.Country.length) {
        return null;
    }

    const country = data.Country[0];

    const {
        flag,
        name,
        population,
        location,
        officialLanguages,
        currencies,
        subregion,
        timezones,
    } = country;

    return (
        <Page title="Countries List">
            <GoBackLink onClick={() => history.back()}>
                <Typography variant="button" color="secondary">
                    {'< Regresar'}
                </Typography>
            </GoBackLink>
            <Typography variant="h1">{country.name}</Typography>
            <CountryCountainer>
                <Grid container>
                    <Grid item sm={12} md={4} lg={3}>
                        <Grid container spacing={2}>
                            <Grid item md={12} lg={12}>
                                <img
                                    src={flag?.svgFile}
                                    alt={`${country.name}-flag`}
                                    height="150"
                                />
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Capital: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{name}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Population: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{population}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Area: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{`${population} km2`}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Latitude: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{location?.latitude}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Longitude: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{location?.longitude}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Official languages: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                {officialLanguages?.map((language, index) => (
                                    <span key={index}>{`${language.name}${
                                        index + 1 < officialLanguages.length
                                            ? ', '
                                            : ''
                                    }`}</span>
                                ))}
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Currencies: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                {currencies?.map((currency, index) => (
                                    <span key={index}>{`${currency.code}${
                                        index + 1 < currencies.length
                                            ? ', '
                                            : ''
                                    }`}</span>
                                ))}
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Region: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{subregion?.region.name}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Subregion: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>{subregion?.name}</span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                <span>Timezones: </span>
                            </Grid>
                            <Grid item md={6} lg={6}>
                                {timezones?.map((timezone, index) => (
                                    <span key={index}>
                                        {' '}
                                        {`${timezone.name}${
                                            index + 1 < timezones.length
                                                ? ', '
                                                : ''
                                        }`}{' '}
                                    </span>
                                ))}
                            </Grid>
                        </Grid>
                    </Grid>
                </Grid>
            </CountryCountainer>
        </Page>
    );
};

export default withRouter(CountryView);
