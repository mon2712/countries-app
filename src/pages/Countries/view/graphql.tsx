import { gql } from '@apollo/client';

import { fragments } from 'pages/Countries/list/fragments';

export const COUNTRY_DATA = gql`
    query GetCountryData($_id: String) {
        Country(_id: $_id) {
            ...CountryAttributes
            area
            subregion {
                _id
                name
                region {
                    _id
                    name
                }
            }
            timezones {
                _id
                name
            }
        }
    }
    ${fragments.country}
`;
