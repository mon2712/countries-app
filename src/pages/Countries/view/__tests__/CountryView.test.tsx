import { MemoryRouter } from 'react-router-dom';
import { create, act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';

import CountryView from '../CountryView';
import { COUNTRY_DATA } from '../graphql';

const mocks: any = [
    {
        request: {
            query: COUNTRY_DATA,
            variables: {
                _id: '3',
            },
        },
        result: {
            data: {
                Country: [
                    {
                        _id: '3',
                        area: 652230,
                        name: 'Afghanistan',
                        capital: 'Kabul',
                        population: 27657145,
                        flag: {
                            _id: '1',
                            svgFile: 'https://restcountries.eu/data/afg.svg',
                        },
                        location: {
                            latitude: 33,
                            longitude: 65,
                        },
                        currencies: [
                            {
                                _id: '25',
                                name: 'Afghan afghani',
                                code: 'AFN',
                                symbol: '؋',
                            },
                        ],
                        officialLanguages: [
                            {
                                _id: '24',
                                name: 'Turkmen',
                                nativeName: 'Türkmen',
                            },
                        ],
                        subregion: {
                            _id: '9',
                            name: 'Southern Asia',
                            region: {
                                _id: '10',
                                name: 'Asia',
                            },
                        },
                        timezones: [
                            {
                                _id: '8',
                                name: 'UTC+04:30',
                            },
                        ],
                    },
                ],
            },
        },
    },
];

describe('CountryView', () => {
    it('while loading', () => {
        let component;

        component = create(
            <MemoryRouter>
                <MockedProvider mocks={mocks} addTypename={false}>
                    <CountryView />
                </MockedProvider>
            </MemoryRouter>,
        );
        const p = component.root.findByType('h1');

        expect(p.children.join('')).toContain('Loading...');
        expect(component.toJSON()).toMatchSnapshot();
    });
});
