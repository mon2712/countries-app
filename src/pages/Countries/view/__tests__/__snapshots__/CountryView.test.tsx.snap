// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryView while loading 1`] = `
<div
  className="sc-hKgILt CUpEl"
>
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Loading...
  </h1>
  <div
    className="sc-eCssSg dsCps"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
